updated MHD EMF compute to use fc-vels returned from the Riemann solver

AstroKriel opened the pull request:
### Description

#### Summary
This PR introduces a new `ComputeEMF_UsingFCVel` implementation that uses the face-centered (FC) velocity field returned directly by the HLLD Riemann solver to compute EMF components, instead of deriving a cell-centered (CC) velocity field from the momentum state variables.

#### Motivation
By operating on the FC velocities already produced by the Riemann solver, we can skip one reconstruction step per dimension, reducing redundant computation (`cc-momentum → cc-velocity → fc-velocity`). This avoids re-deriving quantities that already exist.

#### Limitations
The current approach increases memory usage, as it requires:
- Additional ghost zones for both CC and FC quantities.
- Expanding the Riemann solver domain by 3 cells in each relevant direction.

### Related issues
Are there any GitHub issues that are fixed by this pull request? Add a link to them here.

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [x] I have added a description (see above).
- [ ] I have added a link to any related issues (if applicable; see above).
- [x] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [x] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

AstroKriel committed:
draft compute-emf scheme using fc-vel

AstroKriel committed:
reformat updated scheme (TODO: add ability for diff vel and mag interp schemes)

AstroKriel committed:
update: use the new EMF compute scheme

AstroKriel committed:
add comments to highlight where nghost needs to change for this scheme

BenWibking committed:
SetNumStreams(1) for ComputeEMF

AstroKriel committed:
update comments to clarify changes to num ghost zones

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AstroKriel committed:
fix spelling

AstroKriel committed:
Merge branch 'nkriel-emf-fcvels' of github.com:quokka-astro/quokka into nkriel-emf-fcvels

AstroKriel committed:
resolve clang-tidy suggestions: const
const const

AstroKriel committed:
use different nghost for mhd and non-mhd problems

AstroKriel committed:
fix spelling

AstroKriel committed:
update problem sizes to accommodate having more ghost zones

AstroKriel committed:
remove unused var

AstroKriel committed:
update problem dimension to be multiple of blocking factor

AstroKriel committed:
remove unused nghost_cc in new compute-EMF scheme

BenWibking commented on docs/markdown/gallery.md:
This should go in a separate PR. This change is completely unrelated.

AstroKriel replied:
Oh, whoops, I mass replaced all instances of the typo. Happy to bring this easer egg back and fix it another time.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -164,7 +164,10 @@ template <typename problem_t> class QuokkaSimulation : public AMRSimulation<prob
-	int nghost_vel_ = 2;			// number of ghost cells for face velocity computation (default == 2)
+	// number of ghost cells for face velocity computation (default == 2)
+	// we now need +3 to accommodate the higher-order reconstruction in computeEMF
+	int nghost_vel_ = 2 + (Physics_Traits<problem_t>::is_mhd_enabled ? 3 : 0);
```
I don't think +3 is needed. I think 3 total are needed.

AstroKriel replied:
Ah, I see. Let me check I understood: to interpolate face-centered velocity fields to edge centers in computeEMF with piecewise constant, we need valid cells + 1 ghost zone. If we want higher-order schemes (e.g., PPM), that’s +2 more, so computeEMF ends up needing valid + 3 ghost zones. Those fc-vels come from the Riemann solver, so we need to solve the Riemann problem 3 ghost zones deep. But to get the cc-quantities to the faces for the Riemann solver, we need another 4 ghost zones on top of that: +3 for higher-order interpolation, +1 for slope limiting. So the fc-vels (returned from the Riemann solver) need 3 ghost zones, and the global cc-quantities (that go into the Riemann solver) need 7 ghost zones.

BenWibking replied:
I don't follow this at all.

AstroKriel replied:
I was just working backwards from the domain size needed inside computeEMF, to the size required for the whole simulation. When I originally worked through this with Mark, I concluded that we'd need 3 more ghost zones than what we're now using. So I just wanted to check I understood our choice of ghost zones for global cc- and fc-quantities and the fc-vels returned by the Riemann solver.

BenWibking replied:
Ok. I do not understand the details as you've described them in your chain of reasoning.

Although I don't understand it, I think your chain of reasoning is nonetheless too complicated. There are only two steps required to understand why you can get away with 7 ghost cells:
1. You need 4 ghost cells to get a face velocity that lies on the boundary of a box. This statement is equivalent to saying that you need 4 ghost cells for PPM reconstruction as it's done in Quokka.
2. You need 3 more ghost cells to get 3 face velocities extending perpendicular to one of the sides of the box, since each extra ghost cell yields the flux (and face velocity) for one more layer of cells surrounding the box.
That's it: you need 7 ghost cells.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -1441,8 +1444,9 @@ auto QuokkaSimulation<problem_t>::advanceHydroAtLevel(amrex::MultiFab &state_old
-						 emfAveragingType_);
+		// MHDSystem<problem_t>::ComputeEMF(ec_emf_components_fo, state_old_cc_tmp, state_old_fc_tmp, FOfast_mhd_wavespeeds, emfReconstructionOrder_,
+		// emfAveragingType_);
+		MHDSystem<problem_t>::ComputeEMF_UsingFCVel(ec_emf_components_fo, FOfaceVel, state_old_fc_tmp, FOfast_mhd_wavespeeds, emfReconstructionOrder_);
```
Can you make this a runtime option like the choice of EMF averaging schemes? This can be done within MHDSystem, rather than having to change all of the function calls in QuokkaSimulation.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -1864,8 +1872,13 @@ auto QuokkaSimulation<problem_t>::computeHydroFluxes(amrex::MultiFab const &cons
+	// const int reconstructGhost = 3; // reconstruct *two* additional cells outside valid region
+	// default is 2. we need +1 ghost to get fc-vels in the ghost-zones (for piecewise-constant reconstruction) +3 ghosts to accommodate the higher order
+	// reconstruction we need to do in computeEMF
+	const int reconstructGhost = Physics_Traits<problem_t>::is_mhd_enabled ? 6 : 3;
```
Should be 4 if MHD or 3 otherwise.

BenWibking replied:
Actually, reconstructGhost should just be set to `nghost_vel_+1`.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -2031,7 +2044,11 @@ auto QuokkaSimulation<problem_t>::computeFOHydroFluxes(amrex::MultiFab const &co
+	// const int reconstructRange = 3; // reconstruct *two* additional cells outside valid region
+	// same as above: default is 2. we need +1 ghost to get fc-vels in the ghost-zones (for piecewise-constant reconstruction) +3 ghosts to accommodate the
+	// higher order reconstruction we need to do in computeEMF
+	const int reconstructRange = Physics_Traits<problem_t>::is_mhd_enabled ? 6 : 3;
```
Same as above. 4 if MHD, 3 otherwise.

BenWibking replied:
Just set to `nghost_vel_+1` so this is not duplicated in multiple places.

BenWibking commented on src/simulation.hpp:
```diff
@@ -411,8 +411,13 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
+	// int nghost_fc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 6 : 2; // 6 needed for MHD, otherwise only 2 for tracer particles
+
+	// For our new scheme MHD-scheme, we +3 ghosts to achieve higher order interpolation in computeEMF
+	int nghost_cc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 9 : 6;
```
7 if MHD or 6 otherwise.

BenWibking commented on src/simulation.hpp:
```diff
@@ -411,8 +411,13 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
+
+	// For our new scheme MHD-scheme, we +3 ghosts to achieve higher order interpolation in computeEMF
+	int nghost_cc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 9 : 6;
+	int nghost_fc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 9 : 2;
```
3 if MHD, 2 otherwise.

AstroKriel replied:
Do you mean we need +3 for MHD, or just 3 ghost zones? Because this was 6 ghost zones before I had implemented our updated scheme.

BenWibking replied:
Sorry, this is wrong. This should be 7 with MHD. If there's no MHD, then there are no face-centered vars at all. I forgot that the face-centered ghosts are needed to convert the conserved to primitive vars in all of the ghost cells.

BenWibking commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +334,174 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+	// loop over each box-array on the level
+	// note: all the different centerings still have the same distribution mapping, so it is fine for us to attach our looping to cc FArrayBox
+	// note: cell-centered (cc), face-centered (fc), and edge-centered (ec) data all have a different number of cells
+	for (amrex::MFIter mfi(fcx_mf_cVars[0]); mfi.isValid(); ++mfi) {
```
This needs to use 1 stream. See https://github.com/quokka-astro/quokka/pull/1263

BenWibking replied:
Just to be explicit, the loop should look like this:
```
	constexpr int nstreams = 1; // only run on 1 GPU stream to avoid race conditions
	for (amrex::MFIter mfi(cc_mf_cVars, amrex::MFItInfo().SetNumStreams(nstreams)); mfi.isValid(); ++mfi) {
```

BenWibking commented:
@claude Address the PR review comments.

claude[bot] committed:
Address PR review comments

- Reverted unrelated change to docs/markdown/gallery.md
- Fixed const-correctness warnings in mhd_system.hpp (use ''int const'' consistently)
- Fixed ghost zone calculations (use 3 total for MHD, not +3)
- Implemented runtime option ''mhd.use_fc_velocity_for_emf'' to switch between EMF methods
- Updated reconstructGhost to use nghost_vel_+1 instead of hardcoded values
- Updated nghost_cc_ and nghost_fc_ calculations in simulation.hpp

Note: Stream usage fix (PR #1263 reference) was not implemented as it requires
more context about the specific stream handling pattern needed.

Co-authored-by: Ben Wibking <BenWibking@users.noreply.github.com>

BenWibking commented:
/azp run

BenWibking commented:
Ok, the MHD wave tests pass with the reduced number of ghost zones.

BenWibking commented on inputs/alfven_wave_circular.in:
All of the input files can be reverted, since we only need 7 ghosts.

BenWibking commented:
@claude Revert the changes to the input files.

claude[bot] committed:
Revert changes to input files

As requested, reverting the changes to the problem dimensions in the input files since we only need 7 ghosts.

Co-authored-by: Ben Wibking <BenWibking@users.noreply.github.com>

BenWibking commented:
/azp run

AstroKriel commented:
We should wait for #1210, which fixes the GPU race conditions causing these test failures, to be merged into the main repository, then merge it into this PR to finish up.

BenWibking commented:
> We should wait for #1210, which fixes the GPU race conditions causing these test failures, to be merged into the main repository, then merge it into this PR to finish up.

Which test failures are you referring to?

BenWibking commented:
Also: it would be useful to test whether this method is stable with the non-EP reconstruction methods.

AstroKriel commented:
@BenWibking I see the GPU release build failed to compile. Do you think this is likely a Stromlo-side issue?

BenWibking commented:
> @BenWibking I see the GPU release build failed to compile. Do you think this is likely a Stromlo-side issue?

If you click on it, you can see that it failed to find avatargpu:
```
No agent found in pool avatar which satisfies the specified demands: cmake, Agent.Version -gtVersion 2.163.1
```
So it didn't fail to compile, but it failed to communicate with avatargpu at all.

AstroKriel commented:
> So it didn't fail to compile, but it failed to communicate with avatargpu at all.

Gotcha!

BenWibking commented on src/simulation.hpp:
```diff
@@ -411,8 +411,13 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
+
+	// For our new scheme MHD-scheme, we need 7 ghosts for MHD (4 base + 3 for EMF) or 6 otherwise
+	int nghost_cc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 7 : 6;
+	int nghost_fc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 3 : 2;
```
This was my mistake. `nghost_fc_` should actually be the same as `nghost_cc_` (although we don't use the face-centered vars for anything except MHD now).

AstroKriel replied:
I will fix it tomorrow!

BenWibking replied:
I've fixed it.

BenWibking committed:
update nghost_fc_

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

BenWibking commented:
/azp run

BenWibking committed:
make it safe for GPU

BenWibking committed:
Merge branch 'nkriel-emf-fcvels' of github.com:quokka-astro/quokka into nkriel-emf-fcvels

BenWibking commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +335,175 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+				// define EMF FArrayBox
+				ec_fabs_E_Q[iQuad].resize(box_ec, 1);
+				const auto &E2_Qi = ec_fabs_E_Q[iQuad].array();
+				amrex::ParallelFor(box_ec, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
```
The performance optimization I was going to suggest was to do all four quadrants inside this ParallelFor.

AstroKriel replied:
Done.

BenWibking replied:
Ah ok, now I would try to combine this kernel with the emf averaging kernel. That will let you eliminate the temporary EMF FABs for each quadrant, and instead just store the emf values in local variables.

BenWibking commented:
> We should wait for #1210, which fixes the GPU race conditions causing these test failures, to be merged into the main repository, then merge it into this PR to finish up.

I've merged that PR into this one, and made the necessary change for your new `ComputeEMF_*` function. It should be safe to test on the GPU now.

BenWibking commented:
/azp run

AstroKriel committed:
performance optimisation: compute emf comps on the same kernel

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AstroKriel committed:
fix spelling

AstroKriel committed:
Merge branch 'nkriel-emf-fcvels' of github.com:quokka-astro/quokka into nkriel-emf-fcvels

AstroKriel committed:
keeping clang tidy happy

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

BenWibking commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +335,190 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+			// compute the EMF along the cell-edge using a single kernel (all quadrants inside)
+			{
+				// bind read/write Array4 views on the host (required for GPU lambda capture)
+				amrex::Array4<const double> U0s[4];
```
This is a good example of an error that you can silence, because clang-tidy doesn't understand that std::array<> doesn't really work on GPUs.

BenWibking commented:
@claude Modify the ComputeEMF_UsingFCVel function so that all of the temporary amrex::FArrayBoxes are allocated with amrex::The_Async_Arena(), as described in the AMReX documentation (https://amrex-codes.github.io/amrex/docs_html/GPU.html#async-arena). Also, make sure there are no calls to .resize() for any of the FABs.

This is needed to avoid the FArrayBoxes going out of scope before the GPU kernel runs.

claude[bot] committed:
fix: use amrex::The_Async_Arena() for temporary FArrayBox allocations in ComputeEMF_UsingFCVel

- Replace all .resize() calls with direct construction using async arena
- Prevents FArrayBoxes from going out of scope before GPU kernels run
- Affects ec_fabs_Ui_ieside, ec_fabs_Bi_ieside, and ec_fabs_E_Q arrays

Co-authored-by: Ben Wibking <BenWibking@users.noreply.github.com>

chongchonghe commented:
/azp run

AstroKriel committed:
Merge branch 'development' into nkriel-emf-fcvels

chongchonghe commented:
/azp run

BenWibking commented on src/simulation.hpp:
```diff
@@ -411,8 +411,13 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
-	int nghost_cc_ = 6; // PPM needs nghost >= 3, PPM+flattening needs nghost >= 4, +2 for face velocity ghost cells
-	int nghost_fc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 6 : 2; // 6 needed for MHD, otherwise only 2 for tracer particles
+	// int nghost_cc_ = 6; // PPM needs nghost >= 3, PPM+flattening needs nghost >= 4, +2 for face velocity ghost cells
+	// int nghost_fc_ = Physics_Traits<problem_t>::is_mhd_enabled ? 6 : 2; // 6 needed for MHD, otherwise only 2 for tracer particles
```
These two commented-out lines can be deleted.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -2056,7 +2081,11 @@ auto QuokkaSimulation<problem_t>::computeFOHydroFluxes(amrex::MultiFab const &co
 	const auto dm = dmap[lev];
-	const int reconstructRange = 3; // reconstruct *two* additional cells outside valid region
+
+	// const int reconstructRange = 3; // reconstruct *two* additional cells outside valid region
```
This commented-out line can be deleted.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -1850,8 +1870,13 @@ auto QuokkaSimulation<problem_t>::computeHydroFluxes(amrex::MultiFab const &cons
-	const int reconstructGhost = 3; // reconstruct *two* additional cells outside valid region
-	// we need two additional ghost cells in order to compute two ghost face velocities
+
+	// const int reconstructGhost = 3; // reconstruct *two* additional cells outside valid region
```
This commented-out line can be deleted.

BenWibking commented on src/QuokkaSimulation.hpp:
```diff
@@ -1439,8 +1444,13 @@ auto QuokkaSimulation<problem_t>::advanceHydroAtLevel(amrex::MultiFab &state_old
+		} else {
+			MHDSystem<problem_t>::ComputeEMF(ec_emf_components_fo, state_old_cc_tmp, state_old_fc_tmp, FOfast_mhd_wavespeeds,
+							 emfReconstructionOrder_, emfAveragingType_);
+		}
```
In terms of code maintainability, it would be very useful to move this if statement to a single function in mhd_system.hpp, so it's not duplicated three times in QuokkaSimulation. Then we can easily add new EMF schemes in a single place.

BenWibking reviewed (changes requested):
Minor comments below:

AstroKriel committed:
remove outdated comments

AstroKriel committed:
update input flag to switch emf scheme + centralise call to EMF scheme

AstroKriel committed:
centralise EMF scheme call

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AstroKriel committed:
Merge branch 'development' into nkriel-emf-fcvels

AstroKriel commented:
/azp run
